prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,266 @@
+		withTLS:     tls,
+	}
+}
+
+func fieldsForJob(src jobSource) logrus.Fields {
+	return logrus.Fields{
+		"jobPrefix": src.jobPath(),
+	}
+}
+
+// newGCSJobSource creates a new gcsJobSource from a given bucket and jobPrefix
+func newGCSJobSource(src string) *gcsJobSource {
+	linkPrefix := "gs://"
+	noPrefixSrc := strings.TrimPrefix(src, linkPrefix)
+	if !strings.HasSuffix(noPrefixSrc, "/") { // Cleaning up path
+		noPrefixSrc += "/"
+	}
+	tokens := strings.FieldsFunc(noPrefixSrc, func(c rune) bool { return c == '/' })
+	bucket := tokens[0]
+	buildID := tokens[len(tokens)-1]
+	name := tokens[len(tokens)-2]
+	jobPrefix := strings.TrimPrefix(noPrefixSrc, bucket+"/") // Extra / is not part of prefix, only necessary for URI
+	return &gcsJobSource{
+		source:     src,
+		linkPrefix: linkPrefix,
+		bucket:     bucket,
+		jobPrefix:  jobPrefix,
+		jobName:    name,
+		buildID:    buildID,
+	}
+}
+
+// GCSMarker holds the starting point for the next paginated GCS query
+type GCSMarker struct {
+	XMLName xml.Name `xml:"ListBucketResult"`
+	Marker  string   `xml:"NextMarker"`
+}
+
+// Contents is a single entry returned by the GCS XML API
+type Contents struct {
+	Key string
+}
+
+// GCSReq contains the contents of a GCS XML API list response
+type GCSReq struct {
+	XMLName  xml.Name `xml:"ListBucketResult"`
+	Contents []Contents
+}
+
+// names is a helper function for extracting artifact names in parallel from a GCS XML API response
+func names(content []byte, names chan<- string, src jobSource, wg *sync.WaitGroup) {
+	defer wg.Done()
+	extracted := GCSReq{
+		Contents: []Contents{},
+	}
+	err := xml.Unmarshal(content, &extracted)
+	if err != nil {
+		logrus.WithFields(fieldsForJob(src)).WithError(err).Error("Error unmarshaling artifact names from XML")
+	}
+	for _, c := range extracted.Contents {
+		names <- c.Key
+	}
+}
+
+// tryGCSXMLGet is a helper function that attempts to list the artifacts available for the given job
+// source with the given scheme and number of retries. It returns whatever it was able to get, even if
+// it runs out of retries.
+func tryGCSXMLGet(src jobSource, scheme string, endpoint string, retries int, maxResults int) [][]byte {
+	bucketName, prefix := extractBucketPrefixPair(src.jobPath())
+	bodies := [][]byte{}
+	marker := GCSMarker{}
+	c := http.Client{}
+	if scheme != httpsScheme {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		c = http.Client{Transport: tr}
+		scheme = httpScheme
+	}
+	retry := 0
+	for retry < retries {
+		params := url.Values{}
+		params.Add("prefix", prefix)
+		if marker.Marker != "" {
+			params.Add("marker", marker.Marker)
+		}
+		req := url.URL{
+			Scheme:   scheme,
+			Host:     endpoint,
+			Path:     bucketName,
+			RawQuery: params.Encode(),
+		}
+		resp, err := c.Get(req.String())
+		if err != nil {
+			logrus.WithField("request", req.String()).WithFields(fieldsForJob(src)).WithError(err).Error("Error in GCS XML API GET request")
```

Comment:
Once you create the request, you can create the logger once so you can dedupe these constructions

Reply:
xml call removed in latest